### PR TITLE
Add LossDynamicsAnalyzer for noisy label detection

### PR DESCRIPTION
**Pull request by vinnamkim**
### Summary

- Ticket no. 103075
- Added LossDynamicsAnalyzer for noisy label detection.
- Added Jupyter notebook example for this feature.
- Added `ImagenetWithSubsetDirsExporter` to support this feature.

### How to test
I added some tests for this change.

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [x] I have added the description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md).​
- [ ] I have updated the [documentation](https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly

### License

- [x] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: datumaro/plugins/data_formats/imagenet.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 class ImagenetPath:
     IMAGE_DIR_NO_LABEL = "no_label"
+    SEP_TOKEN = ":"
```
Comment (sooahleex) [inline datumaro/plugins/data_formats/imagenet.py:22]: I could not catch up the reason we need to use this. Why do we apply this kinds of changes? And is it related to LossDynamicsAnalyzer?

Comment (vinnamkim) [reply datumaro/plugins/data_formats/imagenet.py:22]: As you can see in the notebook example, this example requires `ImagenetWithSubsetDirsExporter`. However, the legacy code was implemented to assign the item id by
```python
item_id = osp.join(label, image_name)
```

It makes import, export, and re-import makes impossible. For example,

`Import`
```
train/label_0/img_0.jpg
```
turns into `export`
```
train/label_0/label_0/img_0.jpg
```

We cannot import `train/label_0/label_0` directory structure using the existing code.

File: datumaro/plugins/data_formats/imagenet.py [hunk 2]
```diff
@@ -131,40 +132,56 @@ class ImagenetWithSubsetDirsImporter(ImagenetImporter):
...
+    USE_SUBSET_DIRS = False
 
     def apply(self):
-        def _get_dir_name(id_parts, label_name):
-            if 1 < len(id_parts) and id_parts[0] == label_name:
-                return ""
+        def _get_name(item: DatasetItem) -> str:
+            id_parts = item.id.split(ImagenetPath.SEP_TOKEN)
+
+            if len(id_parts) == 1:
```
Comment (JihwanEom) [inline datumaro/plugins/data_formats/imagenet.py:141]: How about
```suggestion
            import platform

            if platform.system() == "Windows":
```

File: datumaro/components/algorithms/noisy_label_detection/loss_dynamics_analyzer.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    def __init__(self, dataset: IDataset, alpha: float = 0.001) -> None:
+        purpose = dataset.infos().get("purpose")
+        if purpose != "noisy_label_detection":
+            raise DatasetError(
+                f'Dataset infos should have purpose="noisy_label_detection", but it has {purpose}.'
+            )
+
+        task = dataset.infos().get("task")
+        if task != "OTX-MultiClassCls":
```
Comment (JihwanEom) [inline datumaro/components/algorithms/noisy_label_detection/loss_dynamics_analyzer.py:57]: If you are considering multiple tasks, how about to use `if task not in allowed_task_names`?

File: datumaro/components/algorithms/noisy_label_detection/loss_dynamics_analyzer.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            raise DatasetError(
+                f"{task} is not allowed. The allowed task names are {self.allowed_task_names}."
```
Comment (JihwanEom) [inline datumaro/components/algorithms/noisy_label_detection/loss_dynamics_analyzer.py:59]: `self.allowed_task_names` is not defined.

**Review by JihwanEom — CHANGES_REQUESTED** [review]

File: datumaro/components/algorithms/noisy_label_detection/loss_dynamics_analyzer.py (hunk 4 above)
Comment (JihwanEom) [reply datumaro/components/algorithms/noisy_label_detection/loss_dynamics_analyzer.py:59]: Sorry for confusing, it works.

File: datumaro/components/algorithms/noisy_label_detection/loss_dynamics_analyzer.py (hunk 3 above)
Comment (vinnamkim) [reply datumaro/components/algorithms/noisy_label_detection/loss_dynamics_analyzer.py:57]: I'll do in the future if more tasks are added.

File: datumaro/plugins/data_formats/imagenet.py (hunk 2 above)
Comment (vinnamkim) [reply datumaro/plugins/data_formats/imagenet.py:141]: We need to remove `<label>:` prefix regardless of OS types when exporting it.

**Review by JihwanEom — APPROVED** [review]

**Review by sooahleex — APPROVED** [review]

**Pull request merged**